tool/firebase_util.py

```diff
@@ -1,14 +1,67 @@
+import time
 
 import firebase_admin
 from firebase_admin import credentials
 from firebase_admin import firestore
 from katikati_pylib.logging import logging
 
+log = None
+firebase_client = None
+
 def init_firebase_client(CRYPTO_TOKEN_PATH):
-    log = logging.Logger(__file__, CRYPTO_TOKEN_PATH)
+    global log
+    if log is None:
+        log = logging.Logger(__file__, CRYPTO_TOKEN_PATH)
     log.info("Setting up Firebase client")
     firebase_cred = credentials.Certificate(CRYPTO_TOKEN_PATH)
     firebase_admin.initialize_app(firebase_cred)
-    firebase_client = firestore.client()
+    global firebase_client
+    if firebase_client is None:
+        firebase_client = firestore.client()
     log.info("Done")
     return firebase_client
+
+def push_collection_to_firestore(collection_root, documents):
+    log.info (f"push_collection_to_firestore {collection_root}")
+    col = firebase_client.collection(collection_root)
+    time_start = time.perf_counter_ns()
+    for document_dict in documents:
+        firestore_format_document_dict = _convert_to_firestore_format(collection_root, document_dict)
+        push_document_to_firestore(collection_root, firestore_format_document_dict)
+    time_end = time.perf_counter_ns()
+    ms_elapsed = (time_end - time_start) / (1000 * 1000)
+    log.info (f"push_collection_to_firestore {collection_root} in {ms_elapsed} ms")
+
+def push_document_to_firestore(collection_root, data):
+    ref_path = data["__reference_path"]
+    doc_id = data["__id"]
+    sub_collections = data["__subcollections"]
+
+    log.info (f"push_document_to_firestore {ref_path}")
+
+    del data["__reference_path"]
+    del data["__id"]
+    del data["__subcollections"]
+
+    field_data = {}
+    for field in data:
+        if field in sub_collections:
+            # it's not actually a field, it's a collection, process these later
+            continue
+        field_data[field] = data[field]
+    firebase_client.document(ref_path).set(field_data)
+
+    for sub_collection in sub_collections:
+        push_collection_to_firestore(f"{ref_path}/{sub_collection}", data[sub_collection])
+
+def _convert_to_firestore_format(collection_root, data):
+    firestore_format = {}
+    if "__id" in data:
+        firestore_format = data
+    else:
+        doc_id = list(data.keys())[0]
+        firestore_format["__id"] = doc_id
+        firestore_format["__reference_path"] = f"{collection_root}/{doc_id}"
+        firestore_format["__subcollections"] = []
+        firestore_format.update(data[doc_id])
+    return firestore_format
```

Comment:
I don't think this is exactly what I suggested. The idea is to remove the `if "__id" in data` check because it's making a decision that should be made earlier in the process. So this method should be called by `add_to_firebase,py` before calling `push_collection_to_firestore`, and not by `push_collection_to_firestore` itself.

```suggestion
def convert_documents_to_firestore_format(collection_root, documents):
    firestore_documents = []
    for doc in documents:
        firestore_doc = {}
        doc_id = list(doc.keys())[0]
        firestore_doc["__id"] = doc_id
        firestore_doc["__reference_path"] = f"{collection_root}/{doc_id}"
        firestore_doc["__subcollections"] = []
        firestore_doc.update(doc[doc_id])
        firestore_documents.append(firestore_doc)
    return firestore_documents
```

[I haven't tested this code, so please test that it's valid python]